mpisppy/extensions/grad_rho.py

```diff
@@ -0,0 +1,313 @@
+                    ci += 1
+
+        for ndn_i, var in s._mpisppy_data.nonant_indices.items():
+            grads[ndn_i] = pyo.value(self.grad_exprs[s][ndn_i])
+
+        return grads
+
+    def _compute_and_update_rho(self):
+        """ Computes and sets rhos for each scenario and each variable based on scenario dependence of
+        the denominator in rho calculation.
+        """
+
+        opt = self.opt
+        local_scens = opt.local_scenarios.values()
+
+        if self.indep_denom:
+            grad_denom = self._scen_indep_denom()
+            loc_denom = {s: grad_denom for s in local_scens}
+        else:
+            loc_denom = {s: self._scen_dep_denom(s)
+                           for s in opt.local_scenarios.values()}
+
+        costs = {s: self._eval_grad_exprs(s, self.best_xhat_buf.value_array())
+                     for s in opt.local_scenarios.values()}
+
+        local_nodenames = []
+        local_rhos = {}
+        local_rho_mins = {}
+        local_rho_maxes = {}
+        local_rho_wgted_means = {}
+        global_rho_mins = {}
+        global_rho_maxes = {}
+        global_rho_wgted_means = {}
+
+        for k, s in opt.local_scenarios.items():
+            nlens = s._mpisppy_data.nlens
+            for node in s._mpisppy_node_list:
+                if node.name not in local_nodenames:
+                    ndn = node.name
+                    local_nodenames.append(ndn)
+                    nlen = nlens[ndn]
+                    
+                    local_rhos[ndn] = np.zeros(nlen, dtype="d")
+                    local_rho_mins[ndn] = np.zeros(nlen, dtype="d")
+                    local_rho_maxes[ndn] = np.zeros(nlen, dtype="d")
+                    local_rho_wgted_means[ndn] = np.zeros(nlen, dtype="d")
+                    global_rho_mins[ndn] = np.zeros(nlen, dtype="d")
+                    global_rho_maxes[ndn] = np.zeros(nlen, dtype="d")
+                    global_rho_wgted_means[ndn] = np.zeros(nlen, dtype="d")
+
+        for k, s in opt.local_scenarios.items():
+            nlens = s._mpisppy_data.nlens
+            for node in s._mpisppy_node_list:
+                ndn = node.name
+                rhos = local_rhos[ndn]
+                rho_mins = local_rho_mins[ndn]
+                rho_maxes = local_rho_maxes[ndn]
+                rho_wgted_means = local_rho_wgted_means[ndn]
+
+                rhos = np.fromiter(
+                    (
+                        abs(costs[s][ndn, i]/loc_denom[s][ndn, i])
+                        for i, v in enumerate(node.nonant_vardata_list)
+                    ),
+                    dtype="d",
+                    count=nlens[ndn],
+                )
+
+                np.minimum(rho_mins, rhos, out=rho_mins, where=(s._mpisppy_data.prob_coeff[ndn] > 0))
+                np.maximum(rho_maxes, rhos, out=rho_maxes, where=(s._mpisppy_data.prob_coeff[ndn] > 0))
+                rho_wgted_means += s._mpisppy_data.prob_coeff[ndn] * rhos
```

Comment:
Maybe I am missing something, but if not, the use of wgted is confusing here since the only weight I can see being used is the probability and that is needed to compute the mean.